nulib/meadow
# Update code, configs, and tests for AWS developer environment

# Summary 
Update code, configs, and tests for AWS developer environment

# Specific Changes in this PR
- update configs to work with both devstack and cloud9 environments
- use [Hush](https://hexdocs.pm/hush) to load configs from AWS Secrets Manager and Environment
- add Hush transformers to simplify our config `.exs` files
- replace hardcoded bucket names in tests with shared attributes from `bucket_names.ex`
- replace hardcoded index names in front and back end with environment-specific ones
- switch from `Supervisor` to `DynamicSupervisor` to allow phased app start
- add `Repo.wait_for_connection/0` to wait for the database to connect before starting other supervised processes
- make sure container image built from new configs can run on staging
- get all tests passing in all three environments (local/cloud9/github actions)
- refactor LDAP tests to allow dev and test data to live in the same LDAP instance without conflicting
- update Terraform to support SSM configuration
- use ECR image repository for Meadow image instead of Docker Hub

# Version bump required by the PR

See [Semantic Versioning 2.0.0](https://semver.org/) for help discerning which is required.


# Steps to Test

- Try to run this branch locally against the latest devstack. Make sure dev works and tests pass. I've already vetted it for days in Cloud9 (and @bmquinn has had it working there as well), and it's working on staging.
- Look over the docs for [Hush](https://hexdocs.pm/hush) and the `config/dev/test/releases` `.exs` files to understand what's going on there.
- Ask questions, make comments. This is a big PR.

# :rocket: Deployment Notes

**Note** - __if you check any of these boxes go to the [(always open) `main` <- `staging` PR](https://github.com/nulib/meadow/pulls) and add detailed notes and instructions to help out others who may end up deploying your changes to production__

- Backward compatible API changes
- Backwards-incompatible API changes


# Tested/Verified
